### ISSUE TITLE
DefaultTextStyle is confusing, TextStyle's height is useless property, what we want is line spacing

### ISSUE THREAD

**Issue by chris-rutkowski** [flutter/flutter#55546]
Hello,

Please try the following code on beta 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        appBarTheme: AppBarTheme(
          textTheme: TextTheme(
              headline6: TextStyle(
            fontSize: 20,
            height: 3,
          )),
        ),
        textTheme: TextTheme(
          bodyText1: TextStyle(
            fontSize: 20,
            height: 3,
          ),
        ),
      ),
      home: Screen(),
    );
  }
}

class Screen extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: DefaultTextStyle(
            style: Theme.of(context).appBarTheme.textTheme.headline6,
            textHeightBehavior: TextHeightBehavior(
              applyHeightToFirstAscent: false,
            ),
            child: Text('Screen')),
      ),
      // Default Text Style doesn't apply if it is above scaffold
      body: DefaultTextStyle(
        textHeightBehavior: TextHeightBehavior(
          applyHeightToFirstAscent: false,
        ),
        style: TextStyle(),
        child: ListView(
          children: <Widget>[
            Container(
              color: Colors.grey,
              child: Text(
                'line 1\nline 2\nline 3',
                style: Theme.of(context).textTheme.bodyText1,
              ),
            ),
            FlatButton(
              child: Text('next page'),
              onPressed: () => Navigator.of(context).push(
                MaterialPageRoute(builder: (_) => Nested()),
              ),
            )
          ],
        ),
      ),
    );
  }
}

class Nested extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Nested'),
      ),
      body: ListView(
        children: <Widget>[
          Container(
            color: Colors.yellow,
            child: Text(
              'line 1\nline 2\nline 3',
              style: Theme.of(context).textTheme.bodyText1,
            ),
          ),
        ],
      ),
    );
  }
}
```

Result:

<img width="492" alt="Screen Shot 2563-04-24 at 15 50 39" src="https://user-images.githubusercontent.com/58718695/80193630-609c2c80-8643-11ea-90f1-c605ffa6ebff.png">

<img width="492" alt="Screen Shot 2563-04-24 at 15 50 41" src="https://user-images.githubusercontent.com/58718695/80193633-63971d00-8643-11ea-9d04-b096674d8167.png">

My environment

```
C02XN2L5JGH5:host_flutter crutkowski$ flutter --version
Flutter 1.17.0 • channel beta • https://github.com/flutter/flutter.git
Framework • revision d3ed9ec945 (3 weeks ago) • 2020-04-06 14:07:34 -0700
Engine • revision c9506cb8e9
Tools • Dart 2.8.0 (build 2.8.0-dev.18.0 eea9717938)
```

Problems:
- Scaffold overwrites DefaultTextStyle, therefore the DefaultTextStyle widget needs to be nested inside each Scaffold of each screen. If I put DefaultTextStyle above or directly inside MaterialApp it is completely ignored
- AppBar doesn't align correctly text - on the First screen it is displayed higher than it should be.

Proposed solution:
- Remove the height property of TextStyle - I hardly see any use of it
- Remove the textHeightBehavior completly
- Add the line spacing property in TextStyle

Kind regards
Chris

**Comment by HansMuller** [comment]
CC @GaryQian @johnsonmh

**Comment by GaryQian** [comment]
@chris-rutkowski we are aware of the scaffold overriding the DefaultTextStyle.

Regarding your other points, can you please provide what you expect to see instead? It looks like you have simply used to properties to create a strange layout. The results look roughly as expected given the styles you defined.

**Comment by PixelToast** [comment]
Its a multiplier of the font size, to specify an exact line spacing just do `height: 1 + (line spacing / font size)`

**Comment by GaryQian** [comment]
If you are looking for height not affecting the first line, then `TextHeightBehavior(applyHeightToFirstAscent: false)` is exactly what you are looking for. Why do you think this should be removed? It seems to resolve your problem.

As @PixelToast has mentioned, you can use some simple math to achieve exactly what line height you want. The idea of "line spacing" is quite vague actually, and Flutter opts for a more direct and predictable approach to avoid ambiguous lineSpacing which has varied definitions across platforms and even across the same font file.

**Comment by chris-rutkowski** [comment]
@PixelToast I have figured out what it is, your comment is not helpful unfortunately.

@GaryQian I disagree with you, and that’s why I showed the examples.

Please look closely at the “Screen” title on attached screenshots.
The height is set to 3 and `TextHeightBehavior(applyHeightToFirstAscent: false)` is set.
**These two together for single line text should be equivalent of default height 1?** But you can clearly see that text is moved slightly to the top.

And I don’t know what is so vague in line spacing parameter. Do you have pros and cons of research that you did?

**Comment by GaryQian** [comment]
The nested widget has a scaffold which, as I mentioned, is currently overriding the DefaultTextStyle. If you add the TextHeightBehavior to the nested widget, you would see that the space at the top causing misalignment would disappear.

The topic of how fonts define vertical spacing is extremely confusing, but here is the gist of it:

A font file defines a EM square. This is made up of an ascent and descent from the baseline which forms the base units of the font. However, the EM square is not what is used as the height of the font. Instead, there are 3 sets of vertical metrics: hhead, win, and typo. Long story short, these types of metrics are all valid, and derived from when adobe, microsoft, and apple came together and attempted to come up with a font standard. Evidently, nobody wanted to give way. Thus we are stuck with 3 sets of height metrics, and each platform then chooses which of the 3 to display.

In addition, font designers were confused by this, so most fonts are not designed to fit into the EM square as originally intended. Instead, everyone just did what they want, almost always extending beyond the EM square. Thus, when someone is asking for 1.0 line spacing, there is no reliable way to predict how much height is actually used. It can be anywhere from exactly the EM square, to many times taller than the EM square. This variation can even occur within the same font.

Part of the job of Flutter is to deliver a cross platform experience, and with such drastic vertical spacing differences across platforms, designs in one environment may not make any sense compared to designs in others. Thus, we opted to provide a unifying and straightforward way to define heights that would remain consistent across all platforms. Mind you, no functionality is lost, the scaling is just different and synced. You are still fully capable (and generally, with far more precision) control the exact heights of the lines with Flutter's height property compared to "line spacing", which was scaled off of a seemingly arbitrary and variable across platforms line height. We still provide you with this variable line height when height is set to null.

**Comment by chris-rutkowski** [comment]
TLDR, I'm not getting one simple thing here:

If you are able to set the height or all lines (including the first one), why is it so "difficult" to affect all lines but the first? You've already added this behaviour using `DefaultTextStyle`?
All is missing is to make it theme-able in `TextStyle`.

**Comment by GaryQian** [comment]
I don't think you quite understand what line height is in this scenario. `TextHeightBehavior` is a property on the paragraph, which handles the line breaking and layout. `TextStyle` on the other hand only specifies properties for a run of text regardless of how/if that run of text is broken into lines. It does not make sense to put a paragraph level property on a TextStyle, in fact, I have no idea how line properties can even be implemented in a context that has no concept of lines. For example, would it make sense to put a property like `maxLines` on a TextStyle? You may also want to explore `StrutStyle` for additional control of line heights.

It doesn't seem like there are any changes to be made here other than fix the inheritance of DefaultTextStyle in scaffolds, which is already a known issue, so I will close this for now. I am happy to continue help you understand the typography system though!
